proof_of_stake/src/lib.rs

```diff
@@ -5335,6 +5323,198 @@ where
+            let max_below_capacity_validator_amount =
+                get_max_below_capacity_validator_amount(
+                    &below_cap_set,
+                    storage,
+                )?;
+
+            if let Some(max_bc_amount) = max_below_capacity_validator_amount {
+                let below_cap_vals_max =
+                    below_cap_set.at(&max_bc_amount.into());
+                let lowest_position =
+                    find_first_position(&below_cap_vals_max, storage)?.unwrap();
+                let removed_max_below_capacity = below_cap_vals_max
+                    .remove(storage, &lowest_position)?
+                    .expect("Must have been removed");
+
+                insert_validator_into_set(
+                    &consensus_validator_set_handle()
+                        .at(&pipeline_epoch)
+                        .at(&max_bc_amount),
+                    storage,
+                    &pipeline_epoch,
+                    &removed_max_below_capacity,
+                )?;
+                validator_state_handle(&removed_max_below_capacity).set(
+                    storage,
+                    ValidatorState::Consensus,
+                    pipeline_epoch,
+                    0,
+                )?;
+            }
+        }
+        ValidatorState::BelowCapacity => {
+            let below_capacity_set = below_capacity_validator_set_handle()
+                .at(&pipeline_epoch)
+                .at(&pipeline_stake.into());
+            // TODO: handle the unwrap better here
+            let val_position = validator_set_positions_handle()
+                .at(&pipeline_epoch)
+                .get(storage, validator)?
+                .unwrap();
+            let removed = below_capacity_set.remove(storage, &val_position)?;
+            debug_assert_eq!(removed, Some(validator.clone()));
+
+            // Remove position
+            validator_set_positions_handle()
+                .at(&pipeline_epoch)
+                .remove(storage, validator)?;
+        }
+        ValidatorState::BelowThreshold | ValidatorState::Jailed => {}
+        ValidatorState::Inactive => {
+            return Err(DeactivationError::AlreadyInactive(
+                validator.clone(),
+                pipeline_epoch,
+            )
+            .into());
+        }
+    }
+
+    // Set the state to inactive
+    validator_state_handle(validator).set(
+        storage,
+        ValidatorState::Inactive,
+        current_epoch,
+        params.pipeline_len,
+    )?;
+
+    // TODO: what other kinds of data should be deleted, if any??
+
+    Ok(())
+}
+
+/// Re-activate an inactive validator
+pub fn reactivate_validator<S>(
+    storage: &mut S,
+    validator: &Address,
+    current_epoch: Epoch,
+) -> storage_api::Result<()>
+where
+    S: StorageRead + StorageWrite,
+{
+    // TODO: need to additionally check for past slashes, in case a jailed
+    // validator tries to deactivate and then reactivate
```

Comment:
could we prevent deactivation on a jailed validator for this instead?

Reply:
Yeah I suppose this is easier and logical too. Will do.

Reply:
Done in 909afde607e046317958839eb70efbdf457ca1ec